File: Chapter-5/stack_imp_using_LinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.link = None


class Stack_List:
    def __init__(self):
        self.head = 0

    def push_stack(self, key):
        newNode = Node(key)
        newNode.link = self.head
        self.head = newNode

    def pop_stack(self):
        temp = self.head
        if self.head == 0:
            print(" Stack Underflow! ")
        else:
            print(f"Popped element is: {self.head.value}")
            self.head = self.head.link
            del temp

    def peek_stack(self):
        if self.head == 0:
            print(" Stack Underflow! ")
        else:
            print(f"Peek element is: {self.head.value} ")

    def display(self):
        if self.head == 0:
            print(" Stack Underflow! ")
        else:
            print("Elements in the stack are : ")
            temp = self.head
            while temp != 0:
                print(temp.value)
                temp = temp.link


stack = Stack_List()
stack.push_stack(98)
stack.push_stack(56)
stack.push_stack(1)
stack.push_stack(34)
stack.push_stack(20)
stack.display()

stack.pop_stack()
stack.display()

stack.peek_stack()

File: Chapter-5/stack_imp_using_LinkedList.cpp
#include <bits/stdc++.h>

using namespace std;

struct Node{
    int data;
    struct Node *link;
};

struct Node *head=0;

void push_stack(){
    int value;
    cout<<"enter the data: ";
    cin>>value;
    struct Node *new_node = (struct Node*)malloc(sizeof(struct Node));
    new_node->data = value;
    new_node->link = head;
    head = new_node;
}

void pop_stack(){
    struct Node *temp = head;
    if(head == 0){
        cout<<"Underflow!"<<endl;
    }
    else{
        cout<<"popped element is: "<<head->data;
        head = head->link;
        free(temp);
    }
    cout<<endl;
}

void peek_stack(){
    if(head == 0){
        cout<<"Underflow!"<<endl;
    }
    else{
        cout<<"Peek element is : "<<head->data<<endl;
    }
}
void display(){
    struct Node *temp = head;
    if(head == 0){
        cout<<"Underflow!"<<endl;
    }
    else{
        while(temp!=0){
            cout<<temp->data<<" ";
            temp = temp->link;
        }
        cout<<endl;
    }
}

int main(){
    int task;
    do{
        cout<<"1. Push in stack"<<endl;
        cout<<"2. Pop in a stack"<<endl;
        cout<<"3. peak in stack"<<endl;
        cout<<"4. Display The stack"<<endl;
        cout<<"0. For Exit"<<endl;
        cout<<" enter the option : ";
        cin>>task;
        switch(task){
        case 0:
            break;
        case 1:
            push_stack();
            cout<<endl;
            break;

        case 2:
            pop_stack();
            cout<<endl;
            break;

        case 3:
            peek_stack();
            cout<<endl;
            break;

        case 4:
            display();
            cout<<endl;
            break;

        default:
            cout<<"please enter correct option"<<endl;
            cout<<endl;
            break;
        }
    }while(task!=0);

    return 0;
}
